### PR TITLE
Warn when trying many versions

### PR DESCRIPTION
**Pull request by konstin**
In pathological cases, it's hard to follow why the resolution is slow. This is an attempt to give the user an indication what is slow (boto3) and how to fix this (add a manual boto3 constraint). This also helps to determine if this is a pathological case or a server that doesn't support range requests.

I picked the download function as point-of-counting so we can potentially ignore selection versions that we immediately discard, while network requests (both cached and uncached) are what makes the resolution slow. The threasholds i picked are somewhat arbitrary.

```
$ uv pip install bio_embeddings
  ⠸ pytest==8.0.2
warning: Downloading metadata for more than 50 different versions of boto3. Consider adding a stricter version constraint for this package to speed up resolution.
warning: Downloading metadata for more than 50 different versions of botocore. Consider adding a stricter version constraint for this package to speed up resolution.
  ⠼ boto3==1.26.85
```

`warn_user!` interacts badly with the `Reporter`, do we have a way to pause-and-restart the restart the reporter when showing our own messages, similar to tqdm or `IndicatifLayer`?

### CODE REVIEW

**Comment by MichaReiser** [comment]
> warn_user! interacts badly with the Reporter, do we have a way to pause-and-restart the restart the reporter when showing our own messages, similar to tqdm or IndicatifLayer?

Could we lock/buffer the output instead and flush the warnings once the reporter is done?

**Comment by konstin** [comment]
No, we want to report this during the resolution so the user can abort, add constraints and restart. We also want to have during resolution to give a sense for "why is this taking so long".

**Comment by zanieb** [comment]
For what it's worth, I find these warnings kind of _confusing_ in `pip` but I don't have a clear answer why. I think this seems like a good start I guess I'll see if I find these bothersome in the same way :)

**Comment by notatallshaw** [comment]
> For what it's worth, I find these warnings kind of _confusing_ in `pip` but I don't have a clear answer why.

I think I know why it's confusing.

Pip says it is trying many different versions of a package, but there is no clear relationship between the package it names and what a user could do to help solve this problem.

Pip is not saying there are many versions of that package. Pip is not saying that package is the reasons for the slow resolution.

Pip is saying that due to it's algorithm that's what is currently trying a lot. Usually that means Pip has a poor heuristic on what to backtrack on, and it's a sign to someone who wants to improve the resolution algorithm this might be a good requirement example to test on. It is rarely ever helpful to the end user.

I hope once I get my latest PR through that Pip's heuristic on what to backtrack on will be so sufficiently improved the package it mentions in these cases *might* be useful to the end user, but I am not at all confident.

**Comment by konstin** [comment]
What about we merge this and see how users like it? We can roll it back in the next release, it's only a cosmetic change.
